### ISSUE TITLE
Allow classes as template parameter for tvec*, tmat* and tquat.

### ISSUE THREAD

**Issue by patrikhuber** [g-truc/glm#541]
Hi!

I've been trying to use glm together with [Ceres](http://ceres-solver.org), a non-linear solver. Ceres's cost functions are templated on a type T, and at runtime, T will be either double or a custom class called `Jet`, which is used for AutoDiff, but in principle behaves like a double. Now I've tried to use glm functions inside that cost function and in principle it could all work in a pretty awesome way, since all or most of glm's functions are templated on a type T too.

However when trying this, I've encountered the following problem: When instantiating a `glm::tmat4x4` (or any other `tmat`) of type `Jet`, I get an error about attempting to reference a deleted assignment operator in `glm::tvec4`. First the code:

```
using jet_t = ceres::Jet<double, 4>;
glm::tmat4x4<jet_t> glm_mat;
```
And the compile error:
```
1>d:\github\eos\eos\3rdparty\glm-0.9.7.6\glm\detail\type_mat4x4.inl(102): error C2280: 'glm::tvec4<T,0> &glm::tvec4<T,0>::operator =(const glm::tvec4<T,0> &)': attempting to reference a deleted function
1>          with
1>          [
1>              T=jet_t
1>          ]
1>  d:\github\eos\eos\3rdparty\glm-0.9.7.6\glm\detail\type_vec4.hpp(276): note: see declaration of 'glm::tvec4<T,0>::operator ='
1>          with
1>          [
1>              T=jet_t
1>          ]
1>  d:\github\eos\eos\3rdparty\glm-0.9.7.6\glm\detail\type_mat4x4.inl(100): note: while compiling class template member function 'glm::tmat4x4<jet_t,0>::tmat4x4(void)'
1>  D:\Github\eos\eos\examples\eos-v2-ceres.cpp(801): note: see reference to function template instantiation 'glm::tmat4x4<jet_t,0>::tmat4x4(void)' being compiled
1>  D:\Github\eos\eos\examples\eos-v2-ceres.cpp(801): note: see reference to class template instantiation 'glm::tmat4x4<jet_t,0>' being compiled
```

I played around with it but I couldn't really figure out which assignment operator of what class was deleted and why, it seemed like nothing was explicitly marked as deleted in the code.

I've tried `#define GLM_FORCE_NO_CTOR_INIT`, but it doesn't really seem to change anything.

The Ceres guys made their `Jet`-type work with Eigen matrices (for example `Jet` has a default constructor), meaning you can do the same with Eigen, so I'd really hope we could get that work with glm too. If this worked, it would allow using all of glm's functions with Ceres's AutoDiff functionality, which would be pretty awesome!

The source of the `Jet` class can be found [here](https://ceres-solver.googlesource.com/ceres-solver/+/master/include/ceres/jet.h).

/cc @sandwichmaker & @keir, hope that's ok ;-)

**Comment by patrikhuber** [comment]
I dug more into this since it would be really awesome to get it to work. Clang is a bit more explicit in its error message:

```
In file included from ceres_glm_test.cpp:3:
In file included from /home/patrik/glm/glm/mat4x4.hpp:35:
In file included from /home/patrik/glm/glm/detail/type_mat4x4.hpp:243:
/home/patrik/glm/glm/detail/type_mat4x4.inl:102:20: error: object of type 'glm::tvec4<ceres::Jet<double, 4>, glm::precision::highp>' cannot be assigned because its copy assignment operator is
      implicitly deleted
                                this->value[0] = col_type(1, 0, 0, 0);
                                               ^
ceres_glm_test.cpp:11:22: note: in instantiation of member function 'glm::tmat4x4<ceres::Jet<double, 4>, glm::precision::highp>::tmat4x4' requested here
        glm::tmat4x4<jet_t> mmat;
                            ^
/home/patrik/glm/glm/detail/type_vec4.hpp:276:31: note: explicitly defaulted function was implicitly deleted here
                GLM_FUNC_DECL tvec4<T, P> & operator=(tvec4<T, P> const & v) GLM_DEFAULT;
                                            ^
/home/patrik/glm/glm/detail/type_vec4.hpp:139:14: note: copy assignment operator of 'tvec4<ceres::Jet<double, 4>, glm::precision::highp>' is implicitly deleted because variant field 'x' has a
      non-trivial copy assignment operator
                        union { T x, r, s; };

```
(and more of the same).

glm `=default`'s the copy assignment operator of its vec4, in type_vec4.hpp line 276:
`GLM_FUNC_DECL tvec4<T, P> & operator=(tvec4<T, P> const & v) GLM_DEFAULT;`

If I replace this with an "implementation", the error goes away:

```
GLM_FUNC_DECL tvec4<T, P> & operator=(tvec4<T, P> const & v) {
    // Todo: How to implement correctly?
};
```

I was thinking however that the problem should be solvable as well by adding a copy assignment operator to ceres's `Jet` type. But that did not change the error. On the second thought this probably makes sense: No matter how you change the `Jet` type, you cannot make it have a trivial copy assignment operator to "satisfy" glms/the compilers requirements.

So from what I gathered now, I would say (or guess) that a fix would need to happen in glm, and not ceres, but I'm really not sure, I might as well be missing something. Would implementing the above copy assignment operator in glm be a good solution, and if yes how would a correct implementation look like?

**Comment by patrikhuber** [comment]
Okay, I think I got it all figured out now, and it's actually probably a behaviour that we should fix in glm anyway:

In my quest to resolve this, instead of the defaulted copy assignment operator
`GLM_FUNC_DECL tvec4<T, P> & operator=(tvec4<T, P> const & v) GLM_DEFAULT;`
I wrote a custom one, which looked like:
```
GLM_FUNC_DECL tvec4<T, P> & operator=(tvec4<T, P> const & v) {
    this->x = v.x;
    this->x = v.y;
    this->x = v.z;
    this->x = v.w;
    return *this;
};
```
and this seemed to work. After I wrote this, I found that there's actually already a copy constructor _exactly_ like this defined in `type_vec4.inl`, but it's conditioned on `GLM_HAS_DEFAULTED_FUNCTIONS`, i.e. the availability of the C++11 syntax `=default`.

The bottom line is: These two code paths (with and without C++11) were probably meant to behave identically, but in fact they are not, when the type `T` is not trivially copyable.

I think this should be fixed, possibly by removing the `if !GLM_HAS_DEFAULTED_FUNCTIONS`s and not using `=default` at all in this case. I'm happy to prepare a PR, if this all makes sense.

**Comment by patrikhuber** [comment]
After some more thought, I'm pretty confident that this is the proper solution. I've prepared a PR in #543.

**Comment by Groovounet** [comment]
The root problem of you issue is that GLM expects basic types for the template parameter of vec*, mat* and quat* types.

I can picture that's useful and GLM used to support half types (a class doing half to float and float to half conversions all over the place) but I eventually deleted the half type because it cause a lot of issues in GLM for the benefit of an extremely low code...

I think it's possible to address with a bunch of compromises such as: No swizzle operators, no optimization for PODs.

Very likely another case for the need of a define to customize GLM implementation.

Thanks,
Christophe

**Comment by patrikhuber** [comment]
Hmm, my PR seems to work quite nicely. Swizzling is not even something I would expect to be working when giving tvec* a non-POD type. I think the basic operations are enough.
If you stuff in a non-POD, would it silently break on some operations?

But in any case, I don't think this speaks against the PR, because as it is now, a compiler that does not have `=default` would happily allow everything I'm trying to do. I just got unlucky and run into these troubles because my compiler has `=default` basically. All the PR does is to make the behaviour the same for C++ compilers that have `=default` and for those that haven't.

**Comment by keir** [comment]
Hi Patrick,

Cool! Glad to see Jets & Ceres get some more use :) If patches are needed
to Ceres let us know and we'll certainly consider them.

Thanks,
Keir

**Comment by patrikhuber** [comment]
Hi @keir,

Thank you very much for the reply! Much appreciated :-)
My conclusion at this point is that the work needs to be done in glm, and not in Ceres. I don't think it's possible to make `Jet` `is_trivially_copyable` since it has an Eigen matrix as member (and a user-defined constructor). glm is a bit fiddly because of its use of a `union` but the change to fix it and make it work is actually trivial. Let's see if we can get PR #543 or a similar fix into glm.

Cheers,
Patrik

**Comment by Groovounet** [comment]
This issue is resolved in GLM 0.9.8 branch thanks to GLM_FORCE_UNRESTRICTED_GENTYPE define.

Thanks for reporting,
Christophe
